Friendships v1 feature milestone review request

Byabasaija opened the pull request:
In this Milestone, we implemented friendship associations where;
- A user is able to send a friend request
- A user is able to see a list of pending requests sent to him/her
- A user can accept or reject the request sent to them.

What we have implemented in this pull request:
-

**We have created a `Friendship`  model with**
- many to many relationship with user model
- has boolean field `confirmed` to check the confirmation of friendship between user

**We have added methods in `User` model**
 - `#friends` to get all friends list for the user instance.
 - `#pending_friends` to see a list of users pending friends request list.
 - `#friend_request` to get the list of friend request that a user recieved.
 - `#confirm_friend(user)` to confirm the friendship request of user instance with the user on the argument passed.
 - `#reject_friend(user)` to reject the friendship between the users by destroying the friendships data.
 - `#request_received` to get the list of friend who sent a friendship request.
 -  `#request_sent` to get a list of user whom the user sent friendship request to.
 - `#request_sent?(user)` to check the request status for user if it was received from user in the argument.
 - `#request_received?(user)` to check the request status for user if it is sent by them to the user in the argument.
 - `#friends?(user)` to check if the friendship between the user with the user provided in the argument.

When multiple users have been registered to the system.

**How sending friendship works.**
- In all users list in url path `/users` we added a button "invite friendship" 
- This button sends `POST` request to `FriendshipsController#create` method and builds a friendships data between the `current_user` who sent the friendship with the user.

> for eg: if we have two users as 
> 
> current_user to be user1 and this current_user sends request to user2 from above mentioned way.

Now user2 can login with their credentials and see the friendship request from user1 in url path `/friendships`  and with the list of friendships requested listed user can see two buttons as `accept` and `reject` for friendships acceptance and rejection.

**How friendships accept works**
- `accept` button submits a `POST` request to `FriendshipsController#accept` method and with the help of `User#confirm_friend(user)` method the friendship is accepted.

**How friendships reject works**
- `reject` button submits a `POST` request to `FriendshipsController#reject` method and with the help of `User#reject_friend(user)` method the friendship is destroyed from the database.

Any feedback given will be appreciated and highly considered :open_hands:

Byabasaija committed:
Create friendships model and create databasse

Byabasaija committed:
migrate the db

dipbazz committed:
Add user and friendship associations

Add methods #friends, #pending_friends, #friend_requests, #friends?
and #confirm_friend in user model

Byabasaija committed:
Create friendship controller

add method for sending friend request
Add helper for the friendship button

dipbazz committed:
Add friendships accept and request send feature

Create User model method #request_received to check if the request
is received from user and #request_sent to check the request sent
from the user.

Create friendship controller method to accept friendship.

Byabasaija committed:
Add reject friend request functionality.

descholar-ceo commented:
# This Pull Request is under review by [descholar-ceo](https://github.com/descholar-ceo) :eyes: :eyes: 
## Please wait a bit ...

descholar-ceo reviewed (changes requested):
# Status: Changes Required :vertical_traffic_light: :warning: 
## Hello @Byabasaija and @dipbazz  :handshake: :handshake: :handshake: 

:+1: This is good to make it this far, your project is really nice, congratulations on the good work you have done so far.
:+1: I liked that you have tried to implement all of the basic requirements of the project carefully, keep it up, and your project is really almost done.
#### :warning: However we need to make few changes to mark this project approved. See the requested change below :point_down: :point_down: :point_down: 
### Note: All of requested changes which are marked as [Required] are mandatory to be implemented in order to get this project approved, others which are marked as [Optional] are nice to have in your project to make it outstanding

# [Required Changes] :red_circle: 
- [x] User has to accept friendship request to start friendship ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/accept_reject_friendship.png) from Pathwright)
- [x] User can reject a friendship request ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/accept_reject_friendship.png) from Pathwright) => I can't see where to click reject or accept a friendship request
- [x] Posts timeline includes posts written by logged-in user and their friends ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/friends_timeline.png) from Pathwright) => currently you are displaying any post from anyone
- [x] Add rspec for unit test and integration tests

# [Optional changes] :yellow_circle: 
- [x] Add username of a current user in navbar
___
_If anything seems to be unclear to you, reachout and ask anything by commenting on this Pull Request and mention me as [descholar-ceo](https://github.com/descholar-ceo), I will receive a notification and will respond._

[GitHub](https://github.com/descholar-ceo) | [LinkedIn](https://linkedin.com/in/mugirase-emmanuel) | [Twitter](https://twitter.com/descholar3)

_Reviewed by [Mugirase Emmanuel (descholar)](https://github.com/descholar-ceo)_

dipbazz committed:
Update timeline view of post

user can only see the post from their friends and their own
post in the timeline.

Byabasaija committed:
Add all unit tests cases for user model

rindrajosia reviewed (approved):
STATUS: APPROVED :+1: 

**Hi team! Very nice project!**

**You did a great job, approved**

**Go and merge to development.**

`Goodluck in your next Milestone`

Happy coding
Rindra Josia
